oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/oidc/OidcUserInfoEndpointFilter.java

```diff
@@ -0,0 +1,138 @@
+
+/**
+ * A {@code Filter} that processes OpenID User Info requests.
+ *
+ * @author Ido Salomon
+ * @see OidcUserInfo
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-core-1_0.html#UserInfoRequest">5.3.1. UserInfo Request</a>
+ * @since 0.1.1
+ */
+public class OidcUserInfoEndpointFilter extends OncePerRequestFilter {
+
+	/**
+	 * The default endpoint {@code URI} for OpenID User Info requests.
+	 */
+	public static final String DEFAULT_OIDC_USER_INFO_ENDPOINT_URI = "/userinfo";
+
+	private final RequestMatcher requestMatcher;
+	private final OidcUserInfoHttpMessageConverter oidcUserInfoHttpMessageConverter =
+			new OidcUserInfoHttpMessageConverter();
+	private final UserInfoClaimsMapper userInfoClaimsMapper;
+
+	public OidcUserInfoEndpointFilter(UserInfoClaimsMapper userInfoClaimsMapper) {
+		AntPathRequestMatcher userInfoGetMatcher = new AntPathRequestMatcher(
+				DEFAULT_OIDC_USER_INFO_ENDPOINT_URI,
+				HttpMethod.GET.name()
+		);
+		AntPathRequestMatcher userInfoPostMatcher = new AntPathRequestMatcher(
+				DEFAULT_OIDC_USER_INFO_ENDPOINT_URI,
+				HttpMethod.POST.name()
+		);
+		this.requestMatcher = new OrRequestMatcher(userInfoGetMatcher, userInfoPostMatcher);
+		this.userInfoClaimsMapper = userInfoClaimsMapper;
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+			throws ServletException, IOException {
+
+		if (!this.requestMatcher.matches(request)) {
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		Object authenticationDetails = authentication.getDetails();
+		Object principal = authentication.getPrincipal();
+		OidcUserInfo oidcUserInfo = userInfoClaimsMapper.map(principal);
+
+		if (authenticationDetails instanceof OAuth2AccessToken) {
+			oidcUserInfo = getUserInfoClaimsRequestedByScope(oidcUserInfo, ((OAuth2AccessToken) authenticationDetails).getScopes());
+		} else {
+			oidcUserInfo = OidcUserInfo.builder()
+					.subject(oidcUserInfo.getSubject())
+					.build();
+		}
+
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		this.oidcUserInfoHttpMessageConverter.write(
+				oidcUserInfo, MediaType.APPLICATION_JSON, httpResponse);
+	}
+
+	private OidcUserInfo getUserInfoClaimsRequestedByScope(OidcUserInfo userInfo, Set<String> scopes) {
+		Set<String> scopeRequestedClaimNames = getScopeRequestedClaimNames(scopes);
+
+		Map<String, Object> scopeRequestedClaims = userInfo.getClaims().entrySet().stream()
+				.filter(claim -> scopeRequestedClaimNames.contains(claim.getKey()))
+				.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+		return new OidcUserInfo(scopeRequestedClaims);
+	}
+
+	private Set<String> getScopeRequestedClaimNames(Set<String> scopes) {
+		Set<String> scopeRequestedClaimNames = new HashSet<>(Arrays.asList(StandardClaimNames.SUB));
```

Comment:
My IDE warns of `Arrays.asList()` on a single param. Since it's mutable, could change to:

```java
                Set<String> scopeRequestedClaimNames = new HashSet<>();
		scopeRequestedClaimNames.add(StandardClaimNames.SUB);
```